rdrf/rdrf/services/io/reporting/reporting_table.py

```diff
@@ -393,17 +397,25 @@ def _get_sql_alchemy_datatype(self, cde_model):
 
         return self.TYPE_MAP.get(datatype, alc.String)
 
-    def dump_csv(self, stream):
-        import csv
-        writer = csv.writer(stream)
-        select_query = alc.sql.select([self.table])
-        db_connection = self.engine.connect()
-        result = db_connection.execute(select_query)
-        writer.writerow([self.column_labeller.get_label(key)
-                         for key in list(result.keys())])
-        writer.writerows(result)
-        db_connection.close()
-        return stream
+    def csv_generator(self, dump_method):
+        writer = csv.writer(StreamingCSVWriter())
+
+        def database_results():
+            sleep(10)
+            dump_method(self)
+            select_query = alc.sql.select([self.table])
+            db_connection = self.engine.connect()
+            result = db_connection.execute(select_query)
+            yield [self.column_labeller.get_label(key) for key in list(result.keys())]
+
+            for row in result:
+                yield row
+
+            db_connection.close()
+
+        csv_rows = (writer.writerow(row) for row in database_results())
+
+        return itertools.chain(["\n"], csv_rows)
```

Comment:
I'm trying to get it to start returning data back to the client before the `dump_method` is called. You can see that this happens because an empty line is returned, then it sleeps, then the rest of the data is returned.

I can't seem to get it to start responding until a newline is yielded in the generator.

Reply:
Split the CSV generation into two steps:
1. Create the reporting table schema/columns (this should be relatively quick)
2. Stream the reporting rows one by one as they get inserted into the db